include/fiction/algorithms/simulation/sidb/quickexact.hpp

```diff
@@ -104,54 +107,82 @@ class quickexact_impl
+                        charge_lyt_with_assigned_dependent_cell.update_after_charge_change(
+                            dependent_cell_mode::VARIABLE);
 
-                    // If no positively charged SiDB can occur in the layout.
-                    if (!three_state_simulation_required)
-                    {
-                        result.additional_simulation_parameters.emplace_back("base_number", uint64_t{2});
-                        two_state_simulation(charge_lyt_with_assigned_dependent_cell);
+                        conduct_simulation(three_state_simulation_required, charge_lyt_with_assigned_dependent_cell);
                     }
-                    // If positively charged SiDBs can occur in the layout, 3-state simulation is conducted.
                     else
                     {
-                        result.additional_simulation_parameters.emplace_back("base_number", uint64_t{3});
-                        three_state_simulation(charge_lyt_with_assigned_dependent_cell);
+                        charge_distribution_surface charge_lyt_with_assigned_dependent_cell{
+                            static_cast<sidb_cell_clk_lyt_siqad>(layout)};
+                        charge_lyt_with_assigned_dependent_cell.assign_physical_parameters(params.physical_parameters);
+                        charge_lyt_with_assigned_dependent_cell.assign_all_charge_states(sidb_charge_state::NEUTRAL);
+                        charge_lyt_with_assigned_dependent_cell.assign_dependent_cell(
+                            all_sidbs_in_lyt_without_negative_preassigned_ones[0]);
+
+                        charge_lyt_with_assigned_dependent_cell.assign_local_external_potential(
+                            params.local_external_potential);
+                        charge_lyt_with_assigned_dependent_cell.assign_global_external_potential(
+                            params.global_potential);
+
+                        // IMPORTANT: The pre-assigned negatively charged SiDBs (they have to be negatively charged to
+                        // fulfill the population stability) are considered as negatively charged defects in the layout.
+                        // Hence, there are no "real" defects assigned, but in order to set some SiDBs with a fixed
+                        // negative charge, this way of implementation is chosen.
+                        for (const auto& cell : preassigned_negative_sidbs)
+                        {
+                            charge_lyt_with_assigned_dependent_cell.add_sidb_defect_to_potential_landscape(
+                                cell, sidb_defect{sidb_defect_type::UNKNOWN, -1,
+                                                  charge_lyt_with_assigned_dependent_cell.get_phys_params().epsilon_r,
+                                                  charge_lyt_with_assigned_dependent_cell.get_phys_params().lambda_tf});
+                        }
```

Comment:
Code duplication?

Reply:
You are right! I will create one big function.